### PR TITLE
fix: vectors should be demoted to the lowest dimension

### LINKED ISSUE #412: Record type is not correct for downcasting awkward vectors
### Vector Version

1.1.1.post1

### Python Version

3.10.12

### OS / Environment

OS/x

### Describe the bug

```python
import awkward as ak
import vector.backends.awkward

a = ak.zip(
    {
        "x": [10.0, 20.0, 30.0],
        "y": [-10.0, 20.0, 30.0],
        "z": [5.0, 10.0, 15.0],
        "t": [16.0, 31.0, 46.0],
    },
    with_name="Momentum4D",
    behavior=vector.backends.awkward.behavior,
)
b = ak.zip(
    {
        "x": [-10.0, 20.0, -30.0],
        "y": [-10.0, -20.0, 30.0],
        "z": [5.0, -10.0, 15.0],
    },
    with_name="Momentum3D",
    behavior=vector.backends.awkward.behavior,
)
c = ak.zip(
    {"x": [-10.0, 13.0, 15.0], "y": [12.0, -4.0, 41.0]},
    with_name="Momentum2D",
    behavior=vector.backends.awkward.behavior,
)

print("a+b")
print(a+b)
print(ak.parameters(a+b))
print("b+a")
print(b+a)
print(ak.parameters(b+a))
```

returns 
```
a+b
[{x: 0, y: -20, z: 10}, {x: 40, y: 0, z: 0}, {x: 0, y: 60, z: 30}]
{'__record__': 'Momentum4D'}
b+a
[{x: 0, y: -20, z: 10}, {x: 40, y: 0, z: 0}, {x: 0, y: 60, z: 30}]
{'__record__': 'Momentum3D'}
```
So, although the fields available are the ones expected, the record type is incorrect in the first case. Similar issues are there for `a+c` and `b+c`.

### Any additional but relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Saransh-cpp**
## Description

Fixes #412

Output of the snippet attached in the issue:
```
a+b
[{x: 0, y: -20, z: 10}, {x: 40, y: 0, z: 0}, {x: 0, y: 60, z: 30}]
{'__record__': 'Momentum3D'}
b+a
[{x: 0, y: -20, z: 10}, {x: 40, y: 0, z: 0}, {x: 0, y: 60, z: 30}]
{'__record__': 'Momentum3D'}
```

## Checklist

- [ ] Have you followed the guidelines in our Contributing document?
- [ ] Have you checked to ensure there aren't any other open Pull Requests for the required change?
- [ ] Does your submission pass pre-commit? (`$ pre-commit run --all-files` or `$ nox -s lint`)
- [ ] Does your submission pass tests? (`$ pytest` or `$ nox -s tests`)
- [ ] Does the documentation build with your changes? (`$ cd docs; make clean; make html` or `$ nox -s docs`)
- [ ] Does your submission pass the doctests? (`$ xdoctest ./src/vector` or `$ nox -s doctests`)

## Before Merging

- [ ] Summarize the commit messages into a brief review of the Pull request.

### CODE REVIEW

**Comment by Saransh-cpp** [comment]
@nsmith- could you check if this is working well? The new tests and the old tests are passing.

The code is a bit dirty; I will refactor it tomorrow and make this ready for review.

**Comment by Saransh-cpp** [comment]
I think this should be considered a breaking change in vector. @henryiii, how should the version be incremented once this is in? Should it go to `v2.0.0` or just `v1.2.0`?

**Review by jpivarski — APPROVED** [review]
I think this is the right thing to do and I read over the (passing) tests, and I think they cover all of the cases.

**Comment by nsmith-** [comment]
I checked out the branch, things look ok interactively, and the problematic coffea tests pass without the explicit cast now. (though some other tests fail, but I think this is on coffea side... checking)

**Comment by nsmith-** [comment]
Actually no, there is some issue now with the fact that we are subclassing the vector array method mixins:
```
>       boosted = a.boost(-a.boostvec)

tests/test_nanoevents_vector.py:313:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.env/lib/python3.10/site-packages/vector/_methods.py:3578: in boost
    return boost_beta3.dispatch(self, booster)
.env/lib/python3.10/site-packages/vector/_compute/lorentz/boost_beta3.py:393: in dispatch
    return _handler_of(v1, v2)._wrap_result(
.env/lib/python3.10/site-packages/vector/_methods.py:4192: in _handler_of
    handler = _demote_handler_vector(
.env/lib/python3.10/site-packages/vector/_methods.py:4144: in _demote_handler_vector
    if len({_handler_priority.index(obj.__module__) for obj in objects}) != 1:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <tuple_iterator object at 0x12d673e20>

>   if len({_handler_priority.index(obj.__module__) for obj in objects}) != 1:
E   ValueError: 'coffea.nanoevents.methods.vector' is not in list

.env/lib/python3.10/site-packages/vector/_methods.py:4144: ValueError
```

(the `.boostvec` property is just calling `.to_beta3()`)

File: src/vector/backends/awkward.py [hunk 1]
```diff
@@ -577,19 +577,21 @@ def elements(self) -> tuple[ArrayOrRecord]:
...
+            return "Momentum4D" if is_vector else cls.__name__[:-5]
     if issubclass(cls, Vector2D):
-        return "Vector2D"
+        return "Vector2D" if is_vector else cls.__name__[:-5]
     if issubclass(cls, Vector3D):
-        return "Vector3D"
+        return "Vector3D" if is_vector else cls.__name__[:-5]
     if issubclass(cls, Vector4D):
-        return "Vector4D"
+        return "Vector4D" if is_vector else cls.__name__[:-5]
```
Comment (Saransh-cpp) [inline src/vector/backends/awkward.py:594]: This function was always returning Vector classes; hence, the performing vector methods on subclasses inheriting vector mixins did not respect the subclasses -
```python
In [5]: a = ak.zip(
   ...:             {
   ...:                 "pt": [10.0, 20.0, 30.0],
   ...:                 "eta": [0.0, 1.1, 2.2],
   ...:                 "phi": [0.1, 0.9, -1.1],
   ...:                 "mass": [1.0, 1.0, 1.0],
   ...:             },
   ...:             with_name="PtEtaPhiMLorentzVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [6]: a
Out[6]: <PtEtaPhiMLorentzVectorArray [{pt: 10, eta: 0, phi: 0.1, ...}, ...] type='3...'>

In [7]: a + a
Out[7]: <LorentzVectorArray [{x: 19.9, y: 2, z: 0, ...}, ..., {...}] type='3 * Lore...'>

In [8]: a - a
Out[8]: <LorentzVectorArray [{x: 0, y: 0, z: 0, t: 0}, ..., {...}] type='3 * Lorent...'>

In [9]: a.to_Vector4D()
Out[9]: <PtEtaPhiMLorentzVectorArray [{pt: 10, eta: 0, phi: 0.1, ...}, ...] type='3...'>

In [10]: a.to_Vector3D()
Out[10]: <MomentumArray3D [{rho: 10, phi: 0.1, ...}, ..., {...}] type='3 * Momentum3...'>

In [11]: a.to_beta3()
Out[11]: <MomentumArray3D [{rho: 0.995, phi: 0.1, ...}, ...] type='3 * Momentum3D[rh...'>
```

Now, everything works well, but only if a user names the classes as "{anything}Array" -
```python
In [2]: from coffea.nanoevents.methods import vector
   ...: import awkward as ak
   ...: a = ak.zip(
   ...:         {
   ...:             "x": [[1, 2], [], [3], [4]],
   ...:             "y": [[5, 6], [], [7], [8]],
   ...:             "z": [[9, 10], [], [11], [12]],
   ...:             "t": [[50, 51], [], [52], [53]],
   ...:         },
   ...:         with_name="LorentzVector",
   ...:         behavior=vector.behavior,
   ...:     )
   ...: a.boost(-a.boostvec)
Out[2]: <LorentzVectorArray [[{x: 0, y: -8.88e-16, ...}, ...], ...] type='4 * var *...'>

In [3]: a + a
Out[3]: <LorentzVectorArray [[{x: 2, y: 10, z: 18, ...}, ...], ...] type='4 * var *...'>

In [4]: a.to_Vector2D()
Out[4]: <TwoVectorArray [[{x: 1, y: 5}, {...}], ..., [{...}]] type='4 * var * TwoVe...'>

In [5]: a.to_Vector3D()
Out[5]: <ThreeVectorArray [[{x: 1, y: 5, z: 9}, {...}], ...] type='4 * var * ThreeV...'>
```

Maybe I should mention this somewhere in the docs.

**Review by Saransh-cpp — COMMENTED** [review]
@nsmith- could you please look at this again? I tested this locally with coffea's vector deprecation branch and all the tests worked for me.

**Comment by henryiii** [comment]
IMO, I'd be fine with calling this v1.2.0, since to me this is a bug fix - a rather impactful one, but still a bug fix. Before it would make a Momentum4D vector with only three dimensions, now the dimensions line up.

But also feel free to do what seems best to you.

**Comment by lgray** [comment]
Ah - this is great! Just seeing the PR now. :-)

**Comment by nsmith-** [comment]
There is still an issue with `(a-b)` vs. `-(b-a)`:
```python
import awkward as ak
from coffea.nanoevents.methods import vector

a = ak.zip(
    {
        "x": [10.0, 20.0, 30.0],
        "y": [-10.0, 20.0, 30.0],
        "z": [5.0, 10.0, 15.0],
        "t": [16.0, 31.0, 46.0],
    },
    with_name="LorentzVector",
    behavior=vector.behavior,
)
b = ak.zip(
    {
        "x": [-10.0, 20.0, -30.0],
        "y": [-10.0, -20.0, 30.0],
        "z": [5.0, -10.0, 15.0],
    },
    with_name="ThreeVector",
    behavior=vector.behavior,
)
c = ak.zip(
    {"rho": [-10.0, 13.0, 15.0], "phi": [1.22, -1.0, 1.0]},
    with_name="TwoVector",
    behavior=vector.behavior,
)

print("a-b")
print(a-b)
print(ak.parameters(a-b))
print("-(b-a)")
print(-(b-a))
print(ak.parameters(-(b-a)))
```
produces
```
a-b
[{x: 20, y: 0, z: 0, t: 16}, {x: 0, ...}, {x: 60, y: 0, z: 0, t: 46}]
{'__record__': 'LorentzVector'}
-(b-a)
[{x: 20, y: -0, z: -0}, {x: -0, y: 40, z: 20}, {x: 60, y: -0, z: -0}]
{'__record__': 'ThreeVector'}
```
They have a type matching their content, but the type demotion isn't working (there should be no `t` in the first case)
Though `a+b` and `b+a` seem to be working OK.

**Comment by Saransh-cpp** [comment]
I think this is on the coffea side (on the `use_scikithep_vector` branch)? I might be wrong -

https://github.com/CoffeaTeam/coffea/blob/894496fc9fe8d3be79263bf7b03e72c18aaf13de/src/coffea/nanoevents/methods/vector.py#L507-L545

**Comment by nsmith-** [comment]
@Saransh-cpp you're right, it's to do with how we are building the subclasses' behavior dispatch tables. I've made a fix that now passes tests on the coffea PR with this PR.

**Comment by lgray** [comment]
Nice - Since we're deprecating nanoevents.vector in this next release of coffea. I think a reasonable timeline for getting a new version of vector with this is and merging the corresponding coffea PR is the beginning of March. Any faster than that and people may get spooked.

**Comment by lgray** [comment]
@Saransh-cpp concerning the version number for the next release. This isn't a major design-changing reimplementation of the package, but it does change behavior significantly, so `1.2.0` makes since IMO.

**Comment by Saransh-cpp** [comment]
Thanks for the reviews! Given that we don't have any release schedule for vector, I'll just merge this in and create a new release.

**Pull request merged**